mesonbuild/scripts/run_tool.py

```diff
@@ -3,17 +3,86 @@
 
 from __future__ import annotations
 
-import itertools
+import asyncio
+import asyncio.subprocess
 import fnmatch
-import concurrent.futures
+import itertools
+import signal
+import sys
 from pathlib import Path
 
+from .. import mlog
 from ..compilers import lang_suffixes
 from ..mesonlib import quiet_git
+from ..mtest import sh_quote, determine_worker_count, complete_all
 import typing as T
 
-if T.TYPE_CHECKING:
-    import subprocess
+Info = T.TypeVar("Info")
+
+async def run_with_buffered_output(cmdlist: T.List[str]) -> int:
+    """Run the command in cmdlist, buffering the output so that it is
+       not mixed for multiple child processes.  Kill the child on
+       cancellation."""
+    quoted_cmdline = ' '.join(sh_quote(x) for x in cmdlist)
+    p: T.Optional[asyncio.subprocess.Process] = None
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
+        stdo, _ = await p.communicate()
+    except FileNotFoundError as e:
+        print(mlog.blue('>>>'), quoted_cmdline, file=sys.stderr)
+        print(mlog.red('not found:'), e.filename, file=sys.stderr)
+        return 1
+    except asyncio.CancelledError:
+        if p:
+            p.kill()
+            await p.wait()
+            return p.returncode or 1
+        else:
+            return 0
+
+    if stdo:
+        print(mlog.blue('>>>'), quoted_cmdline, flush=True)
+        sys.stdout.buffer.write(stdo)
+    return p.returncode
+
+async def _run_workers(infos: T.Iterable[Info],
+                       fn: T.Callable[[Info], T.Iterable[T.Coroutine[None, None, int]]]) -> int:
+    futures: T.List[asyncio.Future[int]] = []
+    semaphore = asyncio.Semaphore(determine_worker_count())
+
+    async def run_one(worker_coro: T.Coroutine[None, None, int]) -> int:
+        try:
+            async with semaphore:
+                return await worker_coro
+        except asyncio.CancelledError as e:
+            worker_coro.throw(e)
+            return await worker_coro
+
+    def sigterm_handler() -> None:
+        for f in futures:
+            f.cancel()
+
+    if sys.platform != 'win32':
+        loop = asyncio.get_running_loop()
+        loop.add_signal_handler(signal.SIGINT, sigterm_handler)
+        loop.add_signal_handler(signal.SIGTERM, sigterm_handler)
+
+    for i in infos:
+        futures.extend((asyncio.ensure_future(run_one(x)) for x in fn(i)))
+    if not futures:
+        return 0
+
+    try:
+        await complete_all(futures)
```

Comment:
... which leaves us with this as the one obviously useful import from mtest.py, which should potentially be refactored instead?

Reply:
Or moved to utils, but I think a one-off is okay since these are the two places that use asyncio in Meson.